# Functions {#functions}

`valr` has a variety of functions for reading and manipulating intervals. Many of the functions have names similar to their [BEDtools][1] and [Genometricorr][17] counterparts.

Because files from genome analyses can be large and interval comparisons are computationally intensive, many functions were implemented in `Rcpp` for maximum speed. See the [function benchmarks](#fxn-benchmarks) section for timings of individual functions.

Several of the methods in `valr` use [non-standard evaluation (NSE)][4] for an expressive syntax. Columns are referred to as naked, unquoted names and can be used in multiple name/value expressions.



```{r glyph function, echo = FALSE, message = FALSE}
library(valr)
library(dplyr)
library(ggplot2)
library(tibble)

bed_glyph <- function(x, y = NULL, 
                      .fun, ...,
                      x_name = "X",
                      y_name = "Y",
                      start_col = "start",
                      end_col = "end",
                      title_ops = "(x, y)") {
  
  if (!is.null(y)){
    x <- bed_cluster(x)
    x <- group_by(x, .id) 
    x <- mutate(x, bin = row_number(.id),
                title = x_name)
    y <- bed_cluster(y) 
    y <- group_by(y, .id)  
    y <- mutate(y, bin = row_number(.id),
                title = y_name)
    res <- eval(.fun(x, y, ...)) 
    res <- mutate(res, bin = 1, 
                  title = "Result")
    if ("start.x" %in% colnames(res)){
      res <- mutate_(res,
                    start = start_col,
                    end = end_col)
    }
    comb <- bind_rows(x, y, res) 
    comb <- mutate(comb, title = factor(title,
                                  levels = c(x_name,
                                             y_name,
                                             "Result")))
  } else if (is.null(y)){
    x <- bed_cluster(x)
    x <- group_by(x, .id) 
    x <- mutate(x, bin = row_number(.id),
             title = x_name)
    res <- eval(.fun(x, ...)) 
    
    
    res <- mutate(res, bin = 1, 
                   title = "Result")
    
    if ("start.x" %in% colnames(res)){
      res <- mutate_(res,
                    start = start_col,
                    end = end_col)
    }
    
    
    comb <- bind_rows(x, res) 
    comb <- mutate(comb, title = factor(title,
                                  levels = c(x_name,
                                             "Result")))
  }
  p_title = paste0(substitute(.fun), title_ops)
  plot_glyph(comb, p_title)
}


plot_glyph <- function(p_data, p_title){
  ggplot2::ggplot(p_data) + 
    geom_rect(aes(xmin = start, xmax = end,
                  ymin = bin, ymax = bin + 0.9,
                  fill = title), color = "black", alpha = 0.75) + 
    ggtitle(p_title) +
    facet_grid(title ~ ., switch = "y", scales = "free_y", space = "free_y") +
    scale_fill_manual(values = c("#fc8d59", 
                        "#ffffbf", 
                        "#91bfdb")) +
    theme_bw() +
    theme(
      axis.line.y = element_blank(), 
      axis.ticks.y = element_blank(),
      axis.text.y = element_blank(),
      legend.position = "none",
      panel.grid = element_blank(),
      panel.background = element_blank(),
      plot.background = element_blank(),
      panel.border = element_blank()
    )
}
```


```{r NSE, eval = FALSE}
# a `value` column is referred to by name
bed_map(x, y, mean = mean(value), var = var(value))
bed_merge(x, concat = concat(value), max = max(value))
```

## Reading data

`valr` has several methods to read interval data. These methods:

  - Take local files or URLs as input.
  - Return a `data.frame` in `tibble::tbl_df` format.
  - Assign consistent `chrom`, `start` and `end` column names.
  - Use [`readr`][6] for speed.
  - Coerce column types.

The methods include:

  - `read_bed()`: read a BED3+ file
  - `read_bed12()`: read a BED12 file
  - `read_bedgraph()`: read a bedGraph file
  - `read_genome()`: read a UCSC "chrom size" file
  - `read_vcf()`: read the Variant Call Format
  - `read_narrowpeak()`: read narrowPeak files
  - `read_broadpeak()`: read broadPeak files

### BED files 

```{r read_bed}
read_bed(system.file('extdata', '3fields.bed.gz', package = 'valr'))
read_bed(n_fields = 6, system.file('extdata', '6fields.bed.gz', package = 'valr'))
read_bed12(system.file('extdata', '12fields.bed.gz', package = 'valr'))
```

### Genome files

We'll use the `hg19` genome as a reference throughout the examples. 

```{r read_genome}
genome <- read_genome(system.file('extdata', 'hg19.chrom.sizes.gz', package = 'valr'))
genome
```

### bedGraph files

```{r read_bedgraph}
read_bedgraph(system.file('extdata', 'test.bg.gz', package = 'valr'))
```

### VCF files

`read_vcf()` reads VCF files and assigns `chrom`, `start` and `end` columns to be used to downstream interval comparisons. Note the interval size is calculated as the length of the `REF` field in the original file.

```{r read_vcf}
read_vcf(system.file('extdata', 'test.vcf.gz', package = 'valr'))
```
  
## Single set operations

`valr` implements several methods for manipulating sets of intervals. Some methods  operate on a single set of intervals, while others compare two sets of intervals.

Many methods the same name as the corresponding `BEDtool`, and some commonly used `BEDtools` are implemented as `dplyr` pipes (e.g., see the [`group_by`](#groupby) section).

All methods accept one or more sets of `x` and `y` intervals, which must either be created using the [`read` methods](#reading data), or have `chrom`, `start` and `end` columns.

These methods operate on a single set of intervals:

  - `bed_sort()`: order intervals 
  - `bed_cluster()`: Cluster (but don't merge) overlapping/nearby intervals.
  - `bed_complement()`: extract intervals *not* represented by an interval file.
  - `bed_merge()`: combine overlapping and nearby intervals into a single interval.
  - `bed_flank()`: Generate new flanking intervals
  - `bed_slop()`: Expand the size of input intervals
  - `bed_shift()`: Shift the coordinates of an input set, bounded by a genome
  
### Sort 

`bed_sort` orders intervals based on a specification. `is_sorted` asks whether a tbl is already sorted.

```{r sort}
x <- bed_random(genome)
is_sorted(x)

y <- bed_sort(x)
y

is_sorted(y)
```

### Cluster 

`bed_cluster` identifies clustered intervals based on a distance specification and assigns them a unique `.id`.

```{r cluster_glyph, echo = FALSE}
a <- tribble(
  ~chrom, ~start, ~end,
  'chr1',      1,      10,
  'chr1',      15,     20,
  'chr1',      25,     35,
  'chr1',      100,    120
)

bed_glyph(a, .fun = bed_cluster, max_dist = 20, 
          title_ops = "(x, max_dist = 20)") +
  geom_label(aes((end - start) / 2 + start, bin + 0.5, label = .id)) + 
  xlab("") + ylab("")
```

```{r cluster}
x <- bed_random(genome)
y <- bed_cluster(x, max_dist = 1000)
y
```

### Complement 

`bed_complement` identifies intervals in a genome that are not covered by an input.

```{r complement_glyph, echo = FALSE}
a <- tribble(
  ~chrom, ~start, ~end,
  'chr1',      1,      10,
  'chr1',      75,    100    
)
g_genome <- tribble(
    ~chrom, ~size, 
  'chr1',      200
)
bed_glyph(a, .fun = bed_complement, genome = g_genome, title_ops = "(x)")
```

```{r complement}
x <- bed_random(genome)

bed_complement(x, genome)
```

### Merge 

`bed_merge` identifies overlapping intervals and reports new merged ones. `is_merged` asks whether a tbl is already merged. Values from merged intervals can be reported using name / value pairs.

```{r merge_glyph, echo = FALSE}
a <- tribble(
  ~chrom, ~start, ~end,
  'chr1',      1,      50,
  'chr1',      10,     75,
  'chr1',      100,    120
)

bed_glyph(a, .fun = bed_merge, title_ops = "(x)")
```

```{r merge}
n <- 1e6
x <- bed_random(genome, n = n)
is_merged(x)

## add some signal
x <- x %>% mutate(signal = runif(n))

bed_merge(x, maxs = max(signal))
```

### Flank

`bed_flank` creates new intervals that flank -- but do not contain -- the input intervals.

```{r flank_glyph, echo = FALSE}
a <- tribble(
  ~chrom, ~start, ~end,
  'chr1',      25,      50,
  'chr1',      100,     125
)

g_genome = tribble(
  ~chrom, ~size,
  'chr1',      125
)

bed_glyph(a, .fun = bed_flank, genome = g_genome, both = 20, title_ops = "(x, both = 20)")
```

```{r flank, eval = FALSE}
bed_flank(x, genome, both = 100)
```

### Slop

`bed_slop` pads input intervals based on a specification

```{r slop_glyph, echo = FALSE}
a <- tribble(
  ~chrom, ~start, ~end,
  'chr1',      110,      120,
  'chr1',      225,     235
)

g_genome = tribble(
  ~chrom, ~size,
  'chr1',      400
)

bed_glyph(a, .fun = bed_slop, genome = g_genome, both = 20, 
          title_ops = "(x, both = 20)")
```

```{r slop}
bed_slop(x, genome, both = 100)
```

### Shift

`bed_shift` adjusts coordinates toward `start` or `end` by a defined `size`. Intervals created out of bounds are removed, or trimmed.


```{r shift_glyph, echo = FALSE}
a <- tribble(
  ~chrom, ~start, ~end,
  'chr1',      25,      50,
  'chr1',      100,     125
)

g_genome = tribble(
  ~chrom, ~size,
  'chr1',      125
)

bed_glyph(a, .fun = bed_shift, genome = g_genome, size = -20, 
          title_ops = "(x, size = -20)")
```

```{r shift}
bed_shift(x, genome, size = 100)
```

### Interval spacing

Interval spacing is easily computed using `dplyr`. Overlapping intervals must first be merged. Spacing for the first interval of each chromosome is undefined.

```{r spacing}
x <- bed_random(genome)
x <- bed_merge(x)

group_by(x, chrom) %>%
  mutate(.spacing = start - lag(end))
```

## Multiple set operations

These methods compare two sets of intervals:

  - `bed_intersect()`: find overlapping intervals
  - `bed_map()`: apply a function to selected columns for overlapping intervals
  - `bed_subtract()`: Remove intervals based on overlaps between two files
  - `bed_window()`: Find overlapping intervals within a window
  - `bed_closest()`: find the closest intervals independent of overlaps
  
### Intersection

`bed_intersect` is implemented using an [interval tree][16] in `Rcpp`. Column names in the result have `.x` and `.y` suffixes, and an `.overlap` column contains the size of the intersection (values of `0` indicate book-ended, or touching intervals). See the [benchmarks](#benchmarks) section for timing. Though `bed_intersect` is pretty fast already, we intend to further improve upon this by parallization with [`RcppParallel`][15].


```{r intersect_glyph, echo = FALSE}
a <- tribble(
  ~chrom, ~start, ~end,
  'chr1',      25,      50,
  'chr1',      100,     125
)

b <- tribble(
  ~chrom, ~start, ~end,
  'chr1',      30,     75
)

bed_glyph(a, b, .fun = bed_intersect, start_col = "start.x", end_col = "end.x")
```

```{r intersect}
# intersect two sets of 1e6 intervals from hg19
x <- bed_random(genome)
y <- bed_random(genome)

bed_intersect(x, y)

# A records with no overlaps (i.e., `-v`)
bed_intersect(x, y, invert = TRUE)
```

One can achieve behaviour similar to BEDtools by combining `bed_intersect` with `dplyr` tools.

```{r intersect_opts}
# `x` records with overlaps (i.e., `-wa`)
bed_intersect(x, y) %>% select(chrom, start = start.x, end = end.x)

# `y` records with overlaps (i.e., `-wb`)
bed_intersect(x, y) %>% select(chrom, start = start.y, end = end.y)

# Unique records in `x` (i.e., `-u`)
bed_intersect(x, y) %>% select(chrom, start = start.x, end = end.x) %>% unique()

# Count `y` intervals that intersect each `x` interval
bed_intersect(x, y) %>% group_by(chrom, start.x, end.x) %>% summarize(count = n())
```

### Map

`bed_map` maps signals onto intervals. Summary statistics for mapped signals can be specific using NSE with name / value pairs. 

```{r map}
bedfile <- system.file('extdata', 'genes.hg19.chr22.bed.gz', package = 'valr')
bgfile  <- system.file('extdata', 'hela.h3k4.chip.bg.gz', package = 'valr')

x <- read_bed(bedfile, n_fields = 6)
y <- read_bedgraph(bgfile)

bed_map(x, y, means = mean(value.y), sds = sd(value.y))
```

### Subtract

`bed_substract()` removes `x` intervals that intersect with `y`.

```{r subtract_glyph, echo = FALSE}
a <- tribble(
  ~chrom, ~start, ~end,
  'chr1',      1,      100
)

b <- tribble(
  ~chrom, ~start, ~end,
  'chr1',      50,     75
)

bed_glyph(a, b, .fun = bed_subtract)
```

```{r subtract}
x <- bed_random(genome)
y <- bed_random(genome)

bed_subtract(x, y)
```

### Window

`bed_window()` identifies `y` intervals that intersect an expanded window of `x` intervals.

```{r window_glyph, echo = FALSE}
a <- tribble(
  ~chrom, ~start, ~end,
  'chr1',      25,      50,
  'chr1',      100,     125
)

b <- tribble(
  ~chrom, ~start, ~end,
  'chr1',      60,     75
)

g_genome = tribble(
  ~chrom, ~size,
  'chr1',      125
)


bed_glyph(a, b, start_col = "start.x", end_col = "end.x", .fun = bed_window, genome = g_genome, both = 15, 
          title_ops = "(x, y, both = 15)")
```

```{r window, warning = FALSE}
x <- bed_random(genome, n = 100)
y <- bed_random(genome, n = 100)

# a few intersections
bed_intersect(x, y)

# can be expanded by casting a wider net
bed_window(x, y, genome, both = 1e6)
```

### Closest

`bed_closest()` identifies `y` intervals that are closest to `x`.


```{r closest_glyph, echo = FALSE}
a <- tribble(
  ~chrom, ~start, ~end,
  'chr1',      100,     125
)

b <- tribble(
  ~chrom, ~start, ~end,
  'chr1',      25,      50,
  'chr1',     140,     175
)

bed_glyph(a, b, start_col = "start.y", end_col = "end.y", .fun = bed_closest)
```


```{r closest}
x <- bed_random(genome, n = 100)
y <- bed_random(genome, n = 100)

bed_closest(x, y)
```

## Randomzing intervals

`valr` provides methods for creating new random intervals or permutations of existing intervals:

  - `bed_random` generates random intervals from an input `genome`.
  - `bed_shuffle` shuffles coordinates given a set of input intervals.
  - Random sampling of input intervals is done with `dplyr`.

### Random

`bed_random` generates random intervals from an input genome. The numbers of intervals from each `chrom` are proporitional to each chrom `size`.

```{r random}
x <- bed_random(genome, n = 1e6, length = 1e3)
x

# numbers of sampled intervals are proportional to chrom size
group_by(x, chrom) %>% summarize(n = n()) %>% arrange(desc(n))
```

### Sample

Sampling can be done using `dplyr`:

```{r sample}
x <- bed_random(genome)

# sample by number
sample_n(x, 1e3, replace = FALSE)

# or fraction
sample_frac(x, 0.1, replace = FALSE)

# or sample intervals within groups
group_by(x, chrom) %>% sample_n(1)
```

### Shuffle

`bed_shuffle` shuffles input intervals. Interval sizes are equal in the input and output.

```{r shuffle_glyph, echo = FALSE, }
seed = 092316
a <- tribble(
  ~chrom, ~start, ~end,
  'chr1',      0,       20,
  'chr1',      40,      50
)

g_genome <- tribble(
  ~chrom, ~size,
  'chr1',     150
)

bed_glyph(a, .fun = bed_shuffle, genome = g_genome, seed = seed, 
          title_ops = "(x)")
```

```{r shuffle, eval = FALSE}
y <- bed_shuffle(x)

any(x$start == y$start)

all(x$end - x$start == y$end - y$start)
```


## Interval Statistics

`valr` provides several methods to assess statistical properties of interval sets including:

  - `bed_fisher()`: measure overlap significance of two sets of intervals.
  - `bed_absdist()`: quantify absolute distance between query intervals and closest reference intervals
  - `bed_reldist()`: quantify relative distances between query intervals and closest reference intervals
  - `bed_jaccard()`: quantify extent of overlap between two sets of intervals
  - `bed_projection()`: XXX

Several of these methods were described in the [Genometricorr][17] software package.

### Fisher's test

The Fisher's test assesses whether two sets of intervals are drawn from the same background genome.

```{r fisher}
x <- bed_random(genome)
y <- bed_random(genome)

bed_fisher(x, y, genome)
```

### Absolute distance

`bed_absdist()` computes the absolute distance between the midpoint of query intervals and the closest midpoints of a set of reference intervals. Absolute distances are scaled by the inter-reference gap for the chromosome as follows. For `Q` total query points and `R` reference points on a chromosome, scale the distance for each query point `i` to the closest reference point by the inter-reference gap for each chromosome. 

$$
d_i(x,y) = min_{\substack{k}}(|q_i - r_k|)\frac{R}{Length\ of\ chromsome}
$$

By default both absolute and scaled distances are reported as `absdist` and `scaled_absdist` respectively. 

```{r absdist, eval = TRUE}
x <- bed_random(genome)
y <- bed_random(genome)

bed_absdist(x, y, genome)
```

### Relative distance

`bed_reldist()` computes the relative distance between a query interval and the two closest reference intervals. The relative distance can range between `[0,0.5]`. If the overall distribution of distances is biased toward zero, then the query and reference intervals tend to be closer in space ([17]). By default a summary with the frequency of each relative distance is reported.  

```{r reldist, eval = TRUE}
x <- bed_random(genome)
y <- bed_random(genome)

bed_reldist(x, y)
```

### Jaccard similarity

`bed_jaccard()` quantifies the extent of overlap between to sets of intervals. The Jaccard statistic takes values of `[0,1]` and is measured as:

$$
J(x,y) = \frac{\mid x \bigcap y \mid}
              {\mid x \bigcup y \mid} 
       = 
         \frac{\mid x \bigcap y \mid}
              {\mid x \mid + \mid y \mid - \mid x \bigcap y \mid}
$$


```{r jaccard}
x <- bed_random(genome)
y <- bed_random(genome)

bed_jaccard(x, y)
```

## Other topics

### Generating sub-intervals

The `bed_makewindows()` function generates sub-intervals from a list of intervals. This approach is useful for aggregating signals around genomic features in small windows, which can reveal subtle patterns that would be masked by aggregating on entire intervals. This is illustrated by examining the density of [histone marks surrounding transcription start sites][#demo-tss].

```{r bed_makewindows}
x <- read_bed(system.file('extdata', '3fields.bed.gz', package = 'valr'))
bed_makewindows(x, win_size = 10, genome)
```

### Coverting interval formats

The `bed12_to_exons()` function extracts individual exons from a BED12 file.

```{r bed12_to_exons}
bed12_path <- system.file('extdata', 'mm9.bed12.gz', package = 'valr')
x <- read_bed12(bed12_path)
bed12_to_exons(x) 
```

### Grouping related intervals

The `group_by` operation from `dplyr` serves a similar purpose to BEDtools `groupby`, and can be combined with `dplyr::summarize` to calculate summary statistics from groups.

```{r group_by}
x <- bed_random(genome)
y <- bed_random(genome)

bed_intersect(x, y) %>%
  group_by(chrom) %>%
  summarize(overlap.total = sum(.overlap))
```

## Function benchmarks {#fxn-benchmarks}

To illustrate the speed of `valr`, we run the most computationally intensive functions on two sets of 1 million intervals sampled from the `hg19` genome. Many of these operations complete in ~1-2 seconds (note that the build bots on Travis CI are slow and decrease speed somewhat).

```{r func_benchmarks}
x <- bed_random(genome)
y <- bed_random(genome)

library(microbenchmark)
microbenchmark(
  bed_random(genome), bed_closest(x, y),
  bed_intersect(x, y), bed_merge(x),
  bed_subtract(x, y), bed_complement(x, genome),
  times = 1, unit = 's'
)
```

[1]: http://bedtools.readthedocs.org/en/latest/
[2]: https://github.com/arq5x/chrom_sweep
[3]: https://github.com/hadley/dplyr
[4]: https://cran.r-project.org/web/packages/dplyr/vignettes/nse.html
[5]: http://www.rcpp.org/
[6]: https://github.com/hadley/readr
[7]: https://pythonhosted.org/pybedtools/
[8]: http://bedops.readthedocs.org/en/latest/index.html
[9]: https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html
[10]: http://shiny.rstudio.com/
[11]: https://cran.r-project.org/web/packages/bedr/index.html
[12]: https://bioconductor.org/packages/release/bioc/html/IRanges.html
[13]: http://bedtools.readthedocs.io/en/latest/#performance
[14]: https://github.com/jayhesselberth/valr/tree/master/src
[15]: https://rcppcore.github.io/RcppParallel/
[16]: https://github.com/ekg/intervaltree/
[17]: http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529
