# Demonstrations {#demos}

## Creating intron intervals {#demo-intron}

Intron intervals can be identified by calculating:

1. Intervals that are covered by any gene annotation
1. Then, intervals that are not covered by exons.

```{r demo-intron}
# genome <- system.file('extdata', 'hg19.chrom.sizes.gz', package = 'valr')
# 
# exons <-
#   
# introns <- 
# introns
```

## Histone density and transcription start sites {#demo-tss}

This demonstration combines several `valr` functions to analyze the distribution of histone marks near transcription start sites.

```{r demo-tss, warning = FALSE, message = FALSE, fig.align='center', fig.cap='human H3K4me3 signal near transcription starts sites'}
library(valr)
library(dplyr)
library(ggplot2)

bedfile <- system.file('extdata', 'genes.hg19.chr22.bed.gz', package = 'valr')
bgfile  <- system.file('extdata', 'hela.h3k4.chip.bg.gz', package = 'valr')
genomefile <- system.file('extdata', 'hg19.chrom.sizes.gz', package = 'valr')

genes <- read_bed(bedfile, n_fields = 6)
genome <- read_genome(genomefile)

# generate 1 bp intervals, + strand only for now
tss <- genes %>%
  filter(strand == '+') %>%
  mutate(end = start + 1)

region_size <- 1000
win_size <- 50

x <- tss %>%
  bed_slop(genome, both = region_size) %>%
  bed_makewindows(genome, win_size) %>%
  group_by(win_id)

y <- read_bedgraph(bgfile)

res <- bed_map(x, y, sums = sum(value.y)) %>%
  summarize(means = mean(sums), sds = sd(sums))

x_labels <- seq(-region_size, region_size, by = win_size * 5)
x_breaks <- seq(1, 41, by = 5)
sd_limits <- aes(ymax = means + sds, ymin = means - sds)

ggplot(res, aes(x = win_id.x, y = means)) +
  geom_point() + geom_pointrange(sd_limits) + 
  scale_x_continuous(labels = x_labels, breaks = x_breaks) + 
  ggtitle('H3K4me3 ChIP signal near TSSs') +
  xlab('Position\n(bp from TSS)') + ylab('Signal') +
  theme_bw()
```

## Correlations among DNase I hypersensitive sites {#demo-dnase}

Here we use `bed_jaccard` for a large-scale comparison of related datasets. As shown in the [`BEDtools` tutorial](http://quinlanlab.org/tutorials/bedtools/bedtools.html), we can measure the similarity of DNaseI hypersensitivity sites for 20 fetal tissue samples. 

This data was taken from [Maurano *et al.* Systematic Localization of Common Disease-Associated Variation in Regulatory DNA. (2012) *Science*](www.sciencemag.org/content/337/6099/1190.short).

```{r init, warning = FALSE, message = FALSE}
library(valr)
library(dplyr)
library(tidyr)
library(purrr)
library(broom)
library(stringr)
library(ggplot2)
library(ggrepel)
library(ComplexHeatmap)
```

First read all 20 BED files containing DNase I hypersensitivity sites from 20 fetal tissues.

```{r read_files}
dnase_files <- list.files('data/dnasei',pattern = 'merge.bed.gz', full.names = TRUE)
data <- dnase_files %>% map(read_bed, n_fields = 4)
```

Then generate a 20x20 table containing a Jaccard statistic for each of the 400 pairwise comparisons.

```{r jaccard-mat}
res <- data %>%
  cross2(.,.) %>%
  map(lift(bed_jaccard)) %>% 
  map("jaccard") %>%
  flatten_dbl() %>%
  matrix(nrow = 20, ncol = 20)
```

We also need to generate labels for the table from the file names.

```{r jaccard-names}
# names are tissue + sample_num
col_names <- dnase_files %>%
  str_split('/') %>% map(`[[`, 3) %>%
  str_split('\\.') %>% map(`[[`, 1) %>%
  str_split('-') %>% map(`[[`, 1) %>%
  flatten_chr() %>%
  str_replace('^f', '') %>%
  str_c(str_c('-', seq(length(.))))

colnames(res) <- col_names
rownames(res) <- col_names
```

Now the Jaccard coefficients can be visualized in heatmap form.

```{r jaccard-heatmap, fig.cap='Heatmap of Jaccard scores', fig.align='center', fig.width=8}
Heatmap(res, color_space = 'Blues')
```

Finally we can do some PCA analysis on the Jaccard coefficients to identify clusters of related cell types.

```{r jaccard-PCA, fig.cap='PCA analysis of Jaccard scores', fig.align='center', fig.width=8}
pca <- broom::tidy(prcomp(res)) %>% as_data_frame()

pca_comps <- filter(pca, PC <= 2) %>%
  tidyr::spread(PC, value) %>%
  setNames(c('label','PC1','PC2'))

ggplot(pca_comps, aes(PC1, PC2)) +
  geom_point(size = 3, color = 'red') +
  geom_text_repel(aes(label = label))
```
