<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>valr demonstrations</title>
  <meta content="text/html; charset=UTF-8" http-equiv="Content-Type">
  <meta name="description" content="This is a collection of demostrations for the valr package.">
  <meta name="generator" content="bookdown 0.1.7 and GitBook 2.6.7">

  <meta property="og:title" content="valr demonstrations" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This is a collection of demostrations for the valr package." />
  <meta name="github-repo" content="jayhesselberth/valr-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="valr demonstrations" />
  <meta name="twitter:site" content="@jayhesselberth" />
  <meta name="twitter:description" content="This is a collection of demostrations for the valr package." />
  

<meta name="author" content="Jay Hesselberth">

<meta name="date" content="2016-09-12">

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="index.html">
<link rel="next" href="demos.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>


  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">valr demonstrations</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i><b>1.1</b> Installation</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#contribute"><i class="fa fa-check"></i><b>1.2</b> Contribute</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#comparison-to-other-tools"><i class="fa fa-check"></i><b>1.3</b> Comparison to other tools</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="functions.html"><a href="functions.html"><i class="fa fa-check"></i><b>2</b> Functions</a><ul>
<li class="chapter" data-level="2.1" data-path="functions.html"><a href="functions.html#non-standard-evaluation"><i class="fa fa-check"></i><b>2.1</b> Non-standard evaluation</a></li>
<li class="chapter" data-level="2.2" data-path="functions.html"><a href="functions.html#reading-data"><i class="fa fa-check"></i><b>2.2</b> Reading data</a><ul>
<li class="chapter" data-level="2.2.1" data-path="functions.html"><a href="functions.html#bed-files"><i class="fa fa-check"></i><b>2.2.1</b> BED files</a></li>
<li class="chapter" data-level="2.2.2" data-path="functions.html"><a href="functions.html#genome-files"><i class="fa fa-check"></i><b>2.2.2</b> Genome files</a></li>
<li class="chapter" data-level="2.2.3" data-path="functions.html"><a href="functions.html#bedgraph-files"><i class="fa fa-check"></i><b>2.2.3</b> bedGraph files</a></li>
<li class="chapter" data-level="2.2.4" data-path="functions.html"><a href="functions.html#vcf-files"><i class="fa fa-check"></i><b>2.2.4</b> VCF files</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="functions.html"><a href="functions.html#interval-manipulations"><i class="fa fa-check"></i><b>2.3</b> Interval manipulations</a><ul>
<li class="chapter" data-level="2.3.1" data-path="functions.html"><a href="functions.html#single-set-operations"><i class="fa fa-check"></i><b>2.3.1</b> Single set operations</a></li>
<li class="chapter" data-level="2.3.2" data-path="functions.html"><a href="functions.html#multiple-set-operations"><i class="fa fa-check"></i><b>2.3.2</b> Multiple set operations</a></li>
<li class="chapter" data-level="2.3.3" data-path="functions.html"><a href="functions.html#randomzing-intervals"><i class="fa fa-check"></i><b>2.3.3</b> Randomzing intervals</a></li>
<li class="chapter" data-level="2.3.4" data-path="functions.html"><a href="functions.html#other-topics"><i class="fa fa-check"></i><b>2.3.4</b> Other topics</a></li>
<li class="chapter" data-level="2.3.5" data-path="functions.html"><a href="functions.html#interval-statistics"><i class="fa fa-check"></i><b>2.3.5</b> Interval Statistics</a></li>
<li class="chapter" data-level="2.3.6" data-path="functions.html"><a href="functions.html#benchmarks"><i class="fa fa-check"></i><b>2.3.6</b> Benchmarks</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="demos.html"><a href="demos.html"><i class="fa fa-check"></i><b>3</b> Demonstrations</a><ul>
<li class="chapter" data-level="3.1" data-path="demos.html"><a href="demos.html#dnase-i-example."><i class="fa fa-check"></i><b>3.1</b> DNase I Example.</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="functions.html"><a href="functions.html#benchmarks"><i class="fa fa-check"></i><b>4</b> Benchmarks</a><ul>
<li class="chapter" data-level="4.1" data-path="benchmarks.html"><a href="benchmarks.html"><i class="fa fa-check"></i><b>4.1</b> Intersection</a></li>
<li class="chapter" data-level="4.2" data-path="benchmarks.html"><a href="benchmarks.html#closest-1"><i class="fa fa-check"></i><b>4.2</b> Closest</a></li>
<li class="chapter" data-level="4.3" data-path="benchmarks.html"><a href="benchmarks.html#merge-1"><i class="fa fa-check"></i><b>4.3</b> Merge</a></li>
<li class="chapter" data-level="4.4" data-path="benchmarks.html"><a href="benchmarks.html#subtract-1"><i class="fa fa-check"></i><b>4.4</b> Subtract</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">valr demonstrations</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="functions" class="section level1">
<h1><span class="header-section-number">Chapter 2</span> Functions</h1>
<div id="non-standard-evaluation" class="section level2">
<h2><span class="header-section-number">2.1</span> Non-standard evaluation</h2>
<p>Several of the methods in <code>valr</code> use <a href="https://cran.r-project.org/web/packages/dplyr/vignettes/nse.html">non-standard evaluation (NSE)</a> for an expressive syntax. Columns are referred to as naked, unquoted names and can be used in multiple name/value expressions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">bed_map</span>(x, y, <span class="dt">mean =</span> <span class="kw">mean</span>(value), <span class="dt">var =</span> <span class="kw">var</span>(value))
<span class="kw">bed_merge</span>(x, <span class="dt">concat =</span> <span class="kw">concat</span>(value), <span class="dt">max =</span> <span class="kw">max</span>(value))</code></pre></div>
<p><code>valr</code> has a variety of functions for reading and manipulating intervals.</p>
</div>
<div id="reading-data" class="section level2">
<h2><span class="header-section-number">2.2</span> Reading data</h2>
<p><code>valr</code> has several methods to read interval data. These methods:</p>
<ul>
<li>Take local files or URLs as input.</li>
<li>Return a <code>data.frame</code> in <code>tibble::tbl_df</code> format.</li>
<li>Assign consistent <code>chrom</code>, <code>start</code> and <code>end</code> column names.</li>
<li>Use <a href="https://github.com/hadley/readr"><code>readr</code></a> for speed.</li>
<li>Coerce column types.</li>
</ul>
<p>The methods include:</p>
<ul>
<li><code>read_bed()</code>: read a BED3+ file</li>
<li><code>read_bed12()</code>: read a BED12 file</li>
<li><code>read_bedgraph()</code>: read a bedGraph file</li>
<li><code>read_genome()</code>: read a UCSC “chrom size” file</li>
<li><code>read_vcf()</code>: read the Variant Call Format</li>
<li><code>read_narrowpeak()</code>: read narrowPeak files</li>
<li><code>read_broadpeak()</code>: read broadPeak files</li>
</ul>
<div id="bed-files" class="section level3">
<h3><span class="header-section-number">2.2.1</span> BED files</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">read_bed</span>(<span class="kw">system.file</span>(<span class="st">&#39;extdata&#39;</span>, <span class="st">&#39;3fields.bed.gz&#39;</span>, <span class="dt">package =</span> <span class="st">&#39;valr&#39;</span>))
<span class="co">#&gt; # A tibble: 10 × 3</span>
<span class="co">#&gt;   chrom  start    end</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;int&gt;  &lt;int&gt;</span>
<span class="co">#&gt; 1  chr1  11873  14409</span>
<span class="co">#&gt; 2  chr1  14361  19759</span>
<span class="co">#&gt; 3  chr1  14406  29370</span>
<span class="co">#&gt; 4  chr1  34610  36081</span>
<span class="co">#&gt; 5  chr1  69090  70008</span>
<span class="co">#&gt; 6  chr1 134772 140566</span>
<span class="co">#&gt; # ... with 4 more rows</span>
<span class="kw">read_bed</span>(<span class="dt">n_fields =</span> <span class="dv">6</span>, <span class="kw">system.file</span>(<span class="st">&#39;extdata&#39;</span>, <span class="st">&#39;6fields.bed.gz&#39;</span>, <span class="dt">package =</span> <span class="st">&#39;valr&#39;</span>))
<span class="co">#&gt; # A tibble: 10 × 6</span>
<span class="co">#&gt;   chrom  start    end      name score strand</span>
<span class="co">#&gt;   &lt;chr&gt;  &lt;int&gt;  &lt;int&gt;     &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;</span>
<span class="co">#&gt; 1  chr1  11873  14409   DDX11L1     3      +</span>
<span class="co">#&gt; 2  chr1  14361  19759    WASH7P    10      -</span>
<span class="co">#&gt; 3  chr1  14406  29370    WASH7P     7      -</span>
<span class="co">#&gt; 4  chr1  34610  36081   FAM138F     3      -</span>
<span class="co">#&gt; 5  chr1  69090  70008     OR4F5     1      +</span>
<span class="co">#&gt; 6  chr1 134772 140566 LOC729737     3      -</span>
<span class="co">#&gt; # ... with 4 more rows</span>
<span class="kw">read_bed12</span>(<span class="kw">system.file</span>(<span class="st">&#39;extdata&#39;</span>, <span class="st">&#39;12fields.bed.gz&#39;</span>, <span class="dt">package =</span> <span class="st">&#39;valr&#39;</span>))
<span class="co">#&gt; # A tibble: 3 × 12</span>
<span class="co">#&gt;   chrom   start     end      name score strand cds_start cds_end item_rgb</span>
<span class="co">#&gt;   &lt;chr&gt;   &lt;int&gt;   &lt;int&gt;     &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;     &lt;int&gt;   &lt;int&gt;    &lt;chr&gt;</span>
<span class="co">#&gt; 1  chr1 4797973 4836816  testgene     1      +   4797973 4836816        .</span>
<span class="co">#&gt; 2 chr10 4848118 4880877 diffchrom     1      +   4848118 4880877        .</span>
<span class="co">#&gt; 3 chr20 5073253 5152630 negstrand     1      -   5073253 5152630        .</span>
<span class="co">#&gt; # ... with 3 more variables: exon_count &lt;int&gt;, exon_sizes &lt;chr&gt;,</span>
<span class="co">#&gt; #   exon_starts &lt;chr&gt;</span></code></pre></div>
</div>
<div id="genome-files" class="section level3">
<h3><span class="header-section-number">2.2.2</span> Genome files</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genome &lt;-<span class="st"> </span><span class="kw">read_genome</span>(<span class="kw">system.file</span>(<span class="st">&#39;extdata&#39;</span>, <span class="st">&#39;hg19.chrom.sizes.gz&#39;</span>, <span class="dt">package =</span> <span class="st">&#39;valr&#39;</span>))</code></pre></div>
</div>
<div id="bedgraph-files" class="section level3">
<h3><span class="header-section-number">2.2.3</span> bedGraph files</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">read_bedgraph</span>(<span class="kw">system.file</span>(<span class="st">&#39;extdata&#39;</span>, <span class="st">&#39;test.bg.gz&#39;</span>, <span class="dt">package =</span> <span class="st">&#39;valr&#39;</span>))
<span class="co">#&gt; # A tibble: 4 × 4</span>
<span class="co">#&gt;   chrom    start      end value</span>
<span class="co">#&gt;   &lt;chr&gt;    &lt;int&gt;    &lt;int&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1 chr19 49302000 49302300 -1.00</span>
<span class="co">#&gt; 2 chr19 49302300 49302600 -0.75</span>
<span class="co">#&gt; 3 chr19 49302600 49302900 -0.50</span>
<span class="co">#&gt; 4 chr19 49302900 49303200 -0.25</span></code></pre></div>
</div>
<div id="vcf-files" class="section level3">
<h3><span class="header-section-number">2.2.4</span> VCF files</h3>
<p><code>read_vcf()</code> reads VCF files and assigns <code>chrom</code>, <code>start</code> and <code>end</code> columns to be used to downstream interval comparisons. Note the interval size is calculated as the length of the <code>REF</code> field in the original file.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">read_vcf</span>(<span class="kw">system.file</span>(<span class="st">&#39;extdata&#39;</span>, <span class="st">&#39;test.vcf.gz&#39;</span>, <span class="dt">package =</span> <span class="st">&#39;valr&#39;</span>))
<span class="co">#&gt; # A tibble: 115 × 201</span>
<span class="co">#&gt;   CHROM   POS    ID                                          REF   ALT</span>
<span class="co">#&gt;   &lt;int&gt; &lt;int&gt; &lt;chr&gt;                                        &lt;chr&gt; &lt;chr&gt;</span>
<span class="co">#&gt; 1     1 10172     .                                       CCCTAA     C</span>
<span class="co">#&gt; 2     1 10390     . CCCCTAACCCCTAACCCTAACCCTAACCCTAACCCTAACCCTAA     C</span>
<span class="co">#&gt; 3     1 10397     .                                CCCCTAACCCTAA     C</span>
<span class="co">#&gt; 4     1 10403     .       ACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAAC     A</span>
<span class="co">#&gt; 5     1 10409     .             ACCCTAACCCTAACCCTAACCCTAACCCTAAC     A</span>
<span class="co">#&gt; 6     1 10415     .                   ACCCTAACCCTAACCCTAACCCTAAC     A</span>
<span class="co">#&gt; # ... with 109 more rows, and 196 more variables: QUAL &lt;dbl&gt;,</span>
<span class="co">#&gt; #   FILTER &lt;chr&gt;, INFO &lt;chr&gt;, FORMAT &lt;chr&gt;, `101976-101976` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `100920-100920` &lt;chr&gt;, `100231-100231` &lt;chr&gt;, `100232-100232` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `100919-100919` &lt;chr&gt;, `101977-101977` &lt;chr&gt;, `100630-100630` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `100640-100640` &lt;chr&gt;, `100631-100631` &lt;chr&gt;, `101583-101583` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101584-101584` &lt;chr&gt;, `101732-101732` &lt;chr&gt;, `101016-101016` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101708-101708` &lt;chr&gt;, `101730-101730` &lt;chr&gt;, `101582-101582` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101731-101731` &lt;chr&gt;, `101809-101809` &lt;chr&gt;, `101653-101653` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101652-101652` &lt;chr&gt;, `101806-101806` &lt;chr&gt;, `101807-101807` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101808-101808` &lt;chr&gt;, `101810-101810` &lt;chr&gt;, `101811-101811` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101812-101812` &lt;chr&gt;, `101813-101813` &lt;chr&gt;, `101814-101814` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101957-101957` &lt;chr&gt;, `101958-101958` &lt;chr&gt;, `100986-100986` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `100987-100987` &lt;chr&gt;, `101897-101897` &lt;chr&gt;, `102071-102071` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `102106-102106` &lt;chr&gt;, `101040-101040` &lt;chr&gt;, `101167-101167` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101041-101041` &lt;chr&gt;, `101042-101042` &lt;chr&gt;, `101168-101168` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101898-101898` &lt;chr&gt;, `102070-102070` &lt;chr&gt;, `102111-102111` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `100988-100988` &lt;chr&gt;, `101896-101896` &lt;chr&gt;, `102110-102110` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `102620-102620` &lt;chr&gt;, `102621-102621` &lt;chr&gt;, `102947-102947` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `102948-102948` &lt;chr&gt;, `103089-103089` &lt;chr&gt;, `102693-102693` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `102694-102694` &lt;chr&gt;, `102949-102949` &lt;chr&gt;, `102622-102622` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `102623-102623` &lt;chr&gt;, `102624-102624` &lt;chr&gt;, `102722-102722` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `102712-102712` &lt;chr&gt;, `103339-103339` &lt;chr&gt;, `103124-103124` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `103161-103161` &lt;chr&gt;, `103125-103125` &lt;chr&gt;, `103171-103171` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `103338-103338` &lt;chr&gt;, `103372-103372` &lt;chr&gt;, `103193-103193` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `100195-100195` &lt;chr&gt;, `100194-100194` &lt;chr&gt;, `101667-101667` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101137-101137` &lt;chr&gt;, `101291-101291` &lt;chr&gt;, `101292-101292` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101666-101666` &lt;chr&gt;, `32222-32222` &lt;chr&gt;, `32049-32049` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `32050-32050` &lt;chr&gt;, `32411-32411` &lt;chr&gt;, `32221-32221` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `100147-100147` &lt;chr&gt;, `100149-100149` &lt;chr&gt;, `100243-100243` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `100290-100290` &lt;chr&gt;, `100753-100753` &lt;chr&gt;, `100754-100754` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101377-101377` &lt;chr&gt;, `101426-101426` &lt;chr&gt;, `101435-101435` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101523-101523` &lt;chr&gt;, `101877-101877` &lt;chr&gt;, `100043-100043` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `100044-100044` &lt;chr&gt;, `100148-100148` &lt;chr&gt;, `100292-100292` &lt;chr&gt;,</span>
<span class="co">#&gt; #   `101381-101381` &lt;chr&gt;, `101393-101393` &lt;chr&gt;, ...</span></code></pre></div>
</div>
</div>
<div id="interval-manipulations" class="section level2">
<h2><span class="header-section-number">2.3</span> Interval manipulations</h2>
<p><code>valr</code> implements several methods for manipulating sets of intervals. Some methods operate on a single set of intervals, while others compare two sets of intervals.</p>
<p>Many methods the same name as the corresponding <code>BEDtool</code>, and some commonly used <code>BEDtools</code> are implemented as <code>dplyr</code> pipes (e.g., see the <a href="#groupby"><code>group_by</code></a> section).</p>
<p>All methods accept one or more sets of <code>x</code> and <code>y</code> intervals, which must either be created using the <a href="#reading%20data"><code>read</code> methods</a>, or have <code>chrom</code>, <code>start</code> and <code>end</code> columns.</p>
<div id="single-set-operations" class="section level3">
<h3><span class="header-section-number">2.3.1</span> Single set operations</h3>
<p>These methods operate on a single set of intervals:</p>
<ul>
<li><code>bed_sort()</code>: order intervals</li>
<li><code>bed_cluster()</code>: Cluster (but don’t merge) overlapping/nearby intervals.</li>
<li><code>bed_complement()</code>: extract intervals <em>not</em> represented by an interval file.</li>
<li><code>bed_merge()</code>: combine overlapping and nearby intervals into a single interval.</li>
<li><code>bed_flank()</code>: Generate new flanking intervals</li>
<li><code>bed_slop()</code>: Expand the size of input intervals</li>
<li><code>bed_shift()</code>: Shift the coordinates of an input set, bounded by a genome</li>
</ul>
<div id="sort" class="section level4">
<h4><span class="header-section-number">2.3.1.1</span> Sort</h4>
<p><code>bed_sort</code> orders intervals based on a specification. <code>is_sorted</code> asks whether a tbl is already sorted.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome)
<span class="kw">is_sorted</span>(x)
<span class="co">#&gt; [1] FALSE</span>

y &lt;-<span class="st"> </span><span class="kw">bed_sort</span>(x)
y
<span class="co">#&gt; # A tibble: 1,000,000 × 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1  chr1  1118  2118</span>
<span class="co">#&gt; 2  chr1  7737  8737</span>
<span class="co">#&gt; 3  chr1 14104 15104</span>
<span class="co">#&gt; 4  chr1 18658 19658</span>
<span class="co">#&gt; 5  chr1 19850 20850</span>
<span class="co">#&gt; 6  chr1 21419 22419</span>
<span class="co">#&gt; # ... with 1e+06 more rows</span>

<span class="kw">is_sorted</span>(y)
<span class="co">#&gt; [1] TRUE</span></code></pre></div>
</div>
<div id="cluster" class="section level4">
<h4><span class="header-section-number">2.3.1.2</span> Cluster</h4>
<p><code>bed_cluster</code> identifies clustered intervals based on a distance specification and assigns them a unique <code>.id</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome)
y &lt;-<span class="st"> </span><span class="kw">bed_cluster</span>(x, <span class="dt">max_dist =</span> <span class="dv">1000</span>)
y
<span class="co">#&gt; # A tibble: 1,000,000 × 4</span>
<span class="co">#&gt;   chrom     start       end   .id</span>
<span class="co">#&gt;   &lt;chr&gt;     &lt;int&gt;     &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1  chr2 178948946 178949946 57787</span>
<span class="co">#&gt; 2  chr7  78418140  78419140 25450</span>
<span class="co">#&gt; 3  chrY  45020975  45021975 14385</span>
<span class="co">#&gt; 4  chr6 118330682 118331682 38328</span>
<span class="co">#&gt; 5  chr5  35717390  35718390 11465</span>
<span class="co">#&gt; 6  chr8  14300540  14301540  4673</span>
<span class="co">#&gt; # ... with 1e+06 more rows</span></code></pre></div>
</div>
<div id="complement" class="section level4">
<h4><span class="header-section-number">2.3.1.3</span> Complement</h4>
<p><code>bed_complement</code> identifies intervals in a genome that are not covered by an input.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome)

<span class="kw">bed_complement</span>(x, genome)
<span class="co">#&gt; # A tibble: 724,132 × 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;   &lt;fctr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1     1  2679</span>
<span class="co">#&gt; 2   chr1  3679  8089</span>
<span class="co">#&gt; 3   chr1  9089 15343</span>
<span class="co">#&gt; 4   chr1 17445 20050</span>
<span class="co">#&gt; 5   chr1 21050 27278</span>
<span class="co">#&gt; 6   chr1 28335 28448</span>
<span class="co">#&gt; # ... with 7.241e+05 more rows</span></code></pre></div>
</div>
<div id="merge" class="section level4">
<h4><span class="header-section-number">2.3.1.4</span> Merge</h4>
<p><code>bed_merge</code> identifies overlapping intervals and reports new merged ones. <code>is_merged</code> asks whether a tbl is already merged. Values from merged intervals can be reported using name / value pairs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">n &lt;-<span class="st"> </span><span class="fl">1e6</span>
x &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome, <span class="dt">n =</span> n)
<span class="kw">is_merged</span>(x)
<span class="co">#&gt; [1] FALSE</span>

## add some signal
x &lt;-<span class="st"> </span>x %&gt;%<span class="st"> </span><span class="kw">mutate</span>(<span class="dt">signal =</span> <span class="kw">runif</span>(n))

<span class="kw">bed_merge</span>(x, <span class="dt">maxs =</span> <span class="kw">max</span>(signal))
<span class="co">#&gt; # A tibble: 723,414 × 4</span>
<span class="co">#&gt;   chrom start   end    maxs</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;int&gt; &lt;int&gt;   &lt;dbl&gt;</span>
<span class="co">#&gt; 1  chr1  1995  3574 0.61656</span>
<span class="co">#&gt; 2  chr1  5752  6752 0.00495</span>
<span class="co">#&gt; 3  chr1  7414  9964 0.91620</span>
<span class="co">#&gt; 4  chr1 16437 17637 0.81356</span>
<span class="co">#&gt; 5  chr1 20780 21780 0.32964</span>
<span class="co">#&gt; 6  chr1 24767 25767 0.12209</span>
<span class="co">#&gt; # ... with 7.234e+05 more rows</span></code></pre></div>
</div>
<div id="flank" class="section level4">
<h4><span class="header-section-number">2.3.1.5</span> Flank</h4>
<p><code>bed_flank</code> creates new intervals that flank – but do not contain – the input intervals.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">bed_flank</span>(x, genome, <span class="dt">both =</span> <span class="dv">100</span>)</code></pre></div>
</div>
<div id="slop" class="section level4">
<h4><span class="header-section-number">2.3.1.6</span> Slop</h4>
<p><code>bed_slop</code> pads input intervals based on a specification</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">bed_slop</span>(x, genome, <span class="dt">both =</span> <span class="dv">100</span>)</code></pre></div>
</div>
<div id="shift" class="section level4">
<h4><span class="header-section-number">2.3.1.7</span> Shift</h4>
<p><code>bed_shift</code> adjusts coordinates toward <code>start</code> or <code>end</code> by a defined <code>size</code>. Intervals created out of bounds are removed, or trimmed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">bed_shift</span>(x, genome, <span class="dt">size =</span> <span class="dv">100</span>)
<span class="co">#&gt; # A tibble: 1,000,000 × 4</span>
<span class="co">#&gt;   chrom    start      end signal</span>
<span class="co">#&gt;   &lt;chr&gt;    &lt;dbl&gt;    &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt; 1 chr18 5.00e+07 5.00e+07 0.0074</span>
<span class="co">#&gt; 2 chr19 3.95e+07 3.95e+07 0.4664</span>
<span class="co">#&gt; 3  chr4 1.18e+08 1.18e+08 0.4978</span>
<span class="co">#&gt; 4  chr2 4.32e+07 4.32e+07 0.2898</span>
<span class="co">#&gt; 5  chr9 6.83e+07 6.83e+07 0.7329</span>
<span class="co">#&gt; 6  chr3 1.22e+08 1.22e+08 0.7725</span>
<span class="co">#&gt; # ... with 1e+06 more rows</span></code></pre></div>
</div>
<div id="interval-spacing" class="section level4">
<h4><span class="header-section-number">2.3.1.8</span> Interval spacing</h4>
<p>Interval spacing is easily computed using <code>dplyr</code>. Overlapping intervals must first be merged. Spacing for the first interval of each chromosome is undefined.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome)
x &lt;-<span class="st"> </span><span class="kw">bed_merge</span>(x)

x %&gt;%
<span class="st">  </span><span class="kw">group_by</span>(chrom) %&gt;%
<span class="st">  </span><span class="kw">mutate</span>(<span class="dt">.spacing =</span> start -<span class="st"> </span><span class="kw">lag</span>(end))
<span class="co">#&gt; Source: local data frame [724,197 x 4]</span>
<span class="co">#&gt; Groups: chrom [25]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;   chrom start   end .spacing</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;int&gt; &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1  chr1  2040  3040       NA</span>
<span class="co">#&gt; 2  chr1  6963  7963     3923</span>
<span class="co">#&gt; 3  chr1 13766 14766     5803</span>
<span class="co">#&gt; 4  chr1 16192 17514     1426</span>
<span class="co">#&gt; 5  chr1 18080 19190      566</span>
<span class="co">#&gt; 6  chr1 19582 20582      392</span>
<span class="co">#&gt; # ... with 7.242e+05 more rows</span></code></pre></div>
</div>
</div>
<div id="multiple-set-operations" class="section level3">
<h3><span class="header-section-number">2.3.2</span> Multiple set operations</h3>
<p>These methods compare two sets of intervals:</p>
<ul>
<li><code>bed_intersect()</code>: find overlapping intervals</li>
<li><code>bed_map()</code>: apply a function to selected columns for overlapping intervals</li>
<li><code>bed_subtract()</code>: Remove intervals based on overlaps between two files</li>
<li><code>bed_window()</code>: Find overlapping intervals within a window</li>
<li><code>bed_closest()</code>: find the closest intervals independent of overlaps</li>
</ul>
<div id="intersection" class="section level4">
<h4><span class="header-section-number">2.3.2.1</span> Intersection</h4>
<p><code>bed_intersect</code> is implemented using an <a href="https://github.com/ekg/intervaltree/">interval tree</a> in <code>Rcpp</code>. Column names in the result have <code>.x</code> and <code>.y</code> suffixes, and an <code>.overlap</code> column contains the size of the intersection (values of <code>0</code> indicate book-ended, or touching intervals). See the <a href="functions.html#benchmarks">benchmarks</a> section for timing. Though <code>bed_intersect</code> is pretty fast already, we intend to further improve upon this by parallization with <a href="https://rcppcore.github.io/RcppParallel/"><code>RcppParallel</code></a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># intersect two sets of 1e6 intervals from hg19</span>
x &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome)
y &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome)

<span class="kw">bed_intersect</span>(x, y)
<span class="co">#&gt; # A tibble: 646,445 × 6</span>
<span class="co">#&gt;   chrom start.x end.x start.y end.y .overlap</span>
<span class="co">#&gt;   &lt;chr&gt;   &lt;int&gt; &lt;int&gt;   &lt;int&gt; &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1  chr1    3450  4450    2900  3900      450</span>
<span class="co">#&gt; 2  chr1   10287 11287   11286 12286        1</span>
<span class="co">#&gt; 3  chr1   10542 11542   11286 12286      256</span>
<span class="co">#&gt; 4  chr1   10604 11604   11286 12286      318</span>
<span class="co">#&gt; 5  chr1   26866 27866   26937 27937      929</span>
<span class="co">#&gt; 6  chr1   37711 38711   38301 39301      410</span>
<span class="co">#&gt; # ... with 6.464e+05 more rows</span>

<span class="co"># A records with no overlaps (i.e., `-v`)</span>
<span class="kw">bed_intersect</span>(x, y, <span class="dt">invert =</span> <span class="ot">TRUE</span>)
<span class="co">#&gt; Source: local data frame [524,327 x 3]</span>
<span class="co">#&gt; Groups: chrom [25]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;   chrom    start      end</span>
<span class="co">#&gt;   &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1  chrY 59359682 59360682</span>
<span class="co">#&gt; 2  chrY 59343718 59344718</span>
<span class="co">#&gt; 3  chrY 59339334 59340334</span>
<span class="co">#&gt; 4  chrY 59285946 59286946</span>
<span class="co">#&gt; 5  chrY 59280348 59281348</span>
<span class="co">#&gt; 6  chrY 59256175 59257175</span>
<span class="co">#&gt; # ... with 5.243e+05 more rows</span></code></pre></div>
<p>One can achieve behaviour similar to BEDtools by combining <code>bed_intersect</code> with <code>dplyr</code> tools.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># `x` records with overlaps (i.e., `-wa`)</span>
<span class="kw">bed_intersect</span>(x, y) %&gt;%<span class="st"> </span><span class="kw">select</span>(chrom, <span class="dt">start =</span> start.x, <span class="dt">end =</span> end.x)
<span class="co">#&gt; # A tibble: 646,445 × 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1  chr1  3450  4450</span>
<span class="co">#&gt; 2  chr1 10287 11287</span>
<span class="co">#&gt; 3  chr1 10542 11542</span>
<span class="co">#&gt; 4  chr1 10604 11604</span>
<span class="co">#&gt; 5  chr1 26866 27866</span>
<span class="co">#&gt; 6  chr1 37711 38711</span>
<span class="co">#&gt; # ... with 6.464e+05 more rows</span>

<span class="co"># `y` records with overlaps (i.e., `-wb`)</span>
<span class="kw">bed_intersect</span>(x, y) %&gt;%<span class="st"> </span><span class="kw">select</span>(chrom, <span class="dt">start =</span> start.y, <span class="dt">end =</span> end.y)
<span class="co">#&gt; # A tibble: 646,445 × 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1  chr1  2900  3900</span>
<span class="co">#&gt; 2  chr1 11286 12286</span>
<span class="co">#&gt; 3  chr1 11286 12286</span>
<span class="co">#&gt; 4  chr1 11286 12286</span>
<span class="co">#&gt; 5  chr1 26937 27937</span>
<span class="co">#&gt; 6  chr1 38301 39301</span>
<span class="co">#&gt; # ... with 6.464e+05 more rows</span>

<span class="co"># Unique records in `x` (i.e., `-u`)</span>
<span class="kw">bed_intersect</span>(x, y) %&gt;%<span class="st"> </span><span class="kw">select</span>(chrom, <span class="dt">start =</span> start.x, <span class="dt">end =</span> end.x) %&gt;%<span class="st"> </span><span class="kw">unique</span>()
<span class="co">#&gt; # A tibble: 475,588 × 3</span>
<span class="co">#&gt;   chrom start   end</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1  chr1  3450  4450</span>
<span class="co">#&gt; 2  chr1 10287 11287</span>
<span class="co">#&gt; 3  chr1 10542 11542</span>
<span class="co">#&gt; 4  chr1 10604 11604</span>
<span class="co">#&gt; 5  chr1 26866 27866</span>
<span class="co">#&gt; 6  chr1 37711 38711</span>
<span class="co">#&gt; # ... with 4.756e+05 more rows</span>

<span class="co"># Count `y` intervals that intersect each `x` interval</span>
<span class="kw">bed_intersect</span>(x, y) %&gt;%<span class="st"> </span><span class="kw">group_by</span>(chrom, start.x, end.x) %&gt;%<span class="st"> </span><span class="kw">summarize</span>(<span class="dt">count =</span> <span class="kw">n</span>())
<span class="co">#&gt; Source: local data frame [475,588 x 4]</span>
<span class="co">#&gt; Groups: chrom, start.x [?]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;   chrom start.x end.x count</span>
<span class="co">#&gt;   &lt;chr&gt;   &lt;int&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1  chr1    3450  4450     1</span>
<span class="co">#&gt; 2  chr1   10287 11287     1</span>
<span class="co">#&gt; 3  chr1   10542 11542     1</span>
<span class="co">#&gt; 4  chr1   10604 11604     1</span>
<span class="co">#&gt; 5  chr1   26866 27866     1</span>
<span class="co">#&gt; 6  chr1   37711 38711     1</span>
<span class="co">#&gt; # ... with 4.756e+05 more rows</span></code></pre></div>
</div>
<div id="map" class="section level4">
<h4><span class="header-section-number">2.3.2.2</span> Map</h4>
<p><code>bed_map</code> maps signals onto intervals. Summary statistics for mapped signals can be specific using NSE with name / value pairs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">bedfile &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&#39;extdata&#39;</span>, <span class="st">&#39;genes.hg19.chr22.bed.gz&#39;</span>, <span class="dt">package =</span> <span class="st">&#39;valr&#39;</span>)
bgfile  &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">&#39;extdata&#39;</span>, <span class="st">&#39;hela.h3k4.chip.bg.gz&#39;</span>, <span class="dt">package =</span> <span class="st">&#39;valr&#39;</span>)

x &lt;-<span class="st"> </span><span class="kw">read_bed</span>(bedfile, <span class="dt">n_fields =</span> <span class="dv">6</span>)
y &lt;-<span class="st"> </span><span class="kw">read_bedgraph</span>(bgfile)

<span class="kw">bed_map</span>(x, y, <span class="dt">means =</span> <span class="kw">mean</span>(value.y), <span class="dt">sds =</span> <span class="kw">sd</span>(value.y))
<span class="co">#&gt; # A tibble: 591 × 5</span>
<span class="co">#&gt;   chrom  start.x    end.x means   sds</span>
<span class="co">#&gt;   &lt;chr&gt;    &lt;int&gt;    &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt;</span>
<span class="co">#&gt; 1 chr22 16150259 16193004  7.91 7.571</span>
<span class="co">#&gt; 2 chr22 16162065 16172265  1.00    NA</span>
<span class="co">#&gt; 3 chr22 16256331 16287937  1.00 0.000</span>
<span class="co">#&gt; 4 chr22 17071647 17073700  1.00 0.000</span>
<span class="co">#&gt; 5 chr22 17082800 17129720  1.12 0.327</span>
<span class="co">#&gt; 6 chr22 17134598 17156430  1.29 0.588</span>
<span class="co">#&gt; # ... with 585 more rows</span></code></pre></div>
</div>
<div id="subtract" class="section level4">
<h4><span class="header-section-number">2.3.2.3</span> Subtract</h4>
<p><code>bed_substract()</code> removes <code>x</code> intervals that intersect with <code>y</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome)
y &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome)

<span class="kw">bed_subtract</span>(x, y)
<span class="co">#&gt; # A tibble: 723,809 × 3</span>
<span class="co">#&gt;    chrom start   end</span>
<span class="co">#&gt;   &lt;fctr&gt; &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1   chr1 10135 10651</span>
<span class="co">#&gt; 2   chr1 11714 12534</span>
<span class="co">#&gt; 3   chr1 14900 16764</span>
<span class="co">#&gt; 4   chr1 20598 21626</span>
<span class="co">#&gt; 5   chr1 25225 25436</span>
<span class="co">#&gt; 6   chr1 29968 30968</span>
<span class="co">#&gt; # ... with 7.238e+05 more rows</span></code></pre></div>
</div>
<div id="window" class="section level4">
<h4><span class="header-section-number">2.3.2.4</span> Window</h4>
<p><code>bed_window()</code> identifies <code>y</code> intervals that intersect an expanded window of <code>x</code> intervals.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome, <span class="dt">n =</span> <span class="dv">100</span>)
y &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome, <span class="dt">n =</span> <span class="dv">100</span>)

<span class="co"># a few intersections</span>
<span class="kw">bed_intersect</span>(x, y)
<span class="co">#&gt; # A tibble: 0 × 6</span>
<span class="co">#&gt; # ... with 6 variables: chrom &lt;chr&gt;, start.x &lt;int&gt;, end.x &lt;int&gt;,</span>
<span class="co">#&gt; #   start.y &lt;int&gt;, end.y &lt;int&gt;, .overlap &lt;int&gt;</span>

<span class="co"># can be expanded by casting a wider net</span>
<span class="kw">bed_window</span>(x, y, genome, <span class="dt">both =</span> <span class="fl">1e6</span>)
<span class="co">#&gt; # A tibble: 6 × 6</span>
<span class="co">#&gt;   chrom   start.x     end.x   start.y     end.y .overlap</span>
<span class="co">#&gt;   &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1 chr10   5084295   5085295   5179694   5180694     1000</span>
<span class="co">#&gt; 2 chr11   2857611   2858611   2712061   2713061     1000</span>
<span class="co">#&gt; 3 chr13 103725094 103726094 104516400 104517400     1000</span>
<span class="co">#&gt; 4 chr15  38779660  38780660  39050011  39051011     1000</span>
<span class="co">#&gt; 5  chr3 141235271 141236271 141863947 141864947     1000</span>
<span class="co">#&gt; 6  chr8  49949992  49950992  50924834  50925834     1000</span></code></pre></div>
</div>
<div id="closest" class="section level4">
<h4><span class="header-section-number">2.3.2.5</span> Closest</h4>
<p><code>bed_closest()</code> identifies <code>y</code> intervals that are closest to <code>x</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome, <span class="dt">n =</span> <span class="dv">100</span>)
y &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome, <span class="dt">n =</span> <span class="dv">100</span>)

<span class="kw">bed_closest</span>(x, y)
<span class="co">#&gt; # A tibble: 99 × 7</span>
<span class="co">#&gt;   chrom   start.x     end.x   start.y     end.y .overlap .distance</span>
<span class="co">#&gt;   &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt;     &lt;int&gt;    &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1  chr1   4921725   4922725  69226270  69227270        0  64303545</span>
<span class="co">#&gt; 2  chr1  18521748  18522748  69226270  69227270        0  50703522</span>
<span class="co">#&gt; 3  chr1  18739282  18740282  69226270  69227270        0  50485988</span>
<span class="co">#&gt; 4  chr1  56436890  56437890  69226270  69227270        0  12788380</span>
<span class="co">#&gt; 5  chr1 108417647 108418647 136064461 136065461        0  27645814</span>
<span class="co">#&gt; 6  chr1 118151382 118152382 136064461 136065461        0  17912079</span>
<span class="co">#&gt; # ... with 93 more rows</span></code></pre></div>
</div>
</div>
<div id="randomzing-intervals" class="section level3">
<h3><span class="header-section-number">2.3.3</span> Randomzing intervals</h3>
<p><code>valr</code> provides methods for creating new random intervals or permutations of existing intervals:</p>
<ul>
<li><code>bed_random</code> generates random intervals from an input <code>genome</code>.</li>
<li><code>bed_shuffle</code> shuffles coordinates given a set of input intervals.</li>
<li>Random sampling of input intervals is done with <code>dplyr</code>.</li>
</ul>
<div id="random" class="section level4">
<h4><span class="header-section-number">2.3.3.1</span> Random</h4>
<p><code>bed_random</code> generates random intervals from an input genome. The numbers of intervals from each <code>chrom</code> are proporitional to each chrom <code>size</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome, <span class="dt">n =</span> <span class="fl">1e6</span>, <span class="dt">length =</span> <span class="fl">1e3</span>)
x
<span class="co">#&gt; # A tibble: 1,000,000 × 3</span>
<span class="co">#&gt;   chrom     start       end</span>
<span class="co">#&gt;   &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1  chr2 158609707 158610707</span>
<span class="co">#&gt; 2  chrY  40570006  40571006</span>
<span class="co">#&gt; 3  chr7 113761802 113762802</span>
<span class="co">#&gt; 4  chr5  15069746  15070746</span>
<span class="co">#&gt; 5  chr2 193127129 193128129</span>
<span class="co">#&gt; 6 chr19  52909745  52910745</span>
<span class="co">#&gt; # ... with 1e+06 more rows</span>

<span class="co"># numbers of sampled intervals are proportional to chrom size</span>
<span class="kw">group_by</span>(x, chrom) %&gt;%<span class="st"> </span><span class="kw">summarize</span>(<span class="dt">n =</span> <span class="kw">n</span>()) %&gt;%<span class="st"> </span><span class="kw">arrange</span>(<span class="kw">desc</span>(n))
<span class="co">#&gt; # A tibble: 25 × 2</span>
<span class="co">#&gt;   chrom     n</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1  chr1 80424</span>
<span class="co">#&gt; 2  chr2 78429</span>
<span class="co">#&gt; 3  chr3 64212</span>
<span class="co">#&gt; 4  chr4 62340</span>
<span class="co">#&gt; 5  chr5 58001</span>
<span class="co">#&gt; 6  chr6 55008</span>
<span class="co">#&gt; # ... with 19 more rows</span></code></pre></div>
</div>
<div id="sample" class="section level4">
<h4><span class="header-section-number">2.3.3.2</span> Sample</h4>
<p>Sampling can be done using <code>dplyr</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome)

<span class="co"># sample by number</span>
<span class="kw">sample_n</span>(x, <span class="fl">1e3</span>, <span class="dt">replace =</span> <span class="ot">FALSE</span>)
<span class="co">#&gt; # A tibble: 1,000 × 3</span>
<span class="co">#&gt;   chrom     start       end</span>
<span class="co">#&gt;   &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1  chr2 212289263 212290263</span>
<span class="co">#&gt; 2  chr5  86119719  86120719</span>
<span class="co">#&gt; 3  chr8 140585044 140586044</span>
<span class="co">#&gt; 4  chr4  36362470  36363470</span>
<span class="co">#&gt; 5  chr4 147834818 147835818</span>
<span class="co">#&gt; 6  chr9  14044797  14045797</span>
<span class="co">#&gt; # ... with 994 more rows</span>

<span class="co"># or fraction</span>
<span class="kw">sample_frac</span>(x, <span class="fl">0.1</span>, <span class="dt">replace =</span> <span class="ot">FALSE</span>)
<span class="co">#&gt; # A tibble: 100,000 × 3</span>
<span class="co">#&gt;   chrom    start      end</span>
<span class="co">#&gt;   &lt;chr&gt;    &lt;int&gt;    &lt;int&gt;</span>
<span class="co">#&gt; 1  chr5 13154911 13155911</span>
<span class="co">#&gt; 2  chr3 43168654 43169654</span>
<span class="co">#&gt; 3 chr18 61254512 61255512</span>
<span class="co">#&gt; 4  chr6 47738235 47739235</span>
<span class="co">#&gt; 5  chrY 38617492 38618492</span>
<span class="co">#&gt; 6 chr15  3179031  3180031</span>
<span class="co">#&gt; # ... with 9.999e+04 more rows</span>

<span class="co"># or sample intervals within groups</span>
<span class="kw">group_by</span>(x, chrom) %&gt;%<span class="st"> </span><span class="kw">sample_n</span>(<span class="dv">1</span>)
<span class="co">#&gt; Source: local data frame [25 x 3]</span>
<span class="co">#&gt; Groups: chrom [25]</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;   chrom     start       end</span>
<span class="co">#&gt;   &lt;chr&gt;     &lt;int&gt;     &lt;int&gt;</span>
<span class="co">#&gt; 1  chr1 105747789 105748789</span>
<span class="co">#&gt; 2 chr10  49907928  49908928</span>
<span class="co">#&gt; 3 chr11  62787101  62788101</span>
<span class="co">#&gt; 4 chr12  45964817  45965817</span>
<span class="co">#&gt; 5 chr13  60548688  60549688</span>
<span class="co">#&gt; 6 chr14  34531511  34532511</span>
<span class="co">#&gt; # ... with 19 more rows</span></code></pre></div>
</div>
<div id="shuffle" class="section level4">
<h4><span class="header-section-number">2.3.3.3</span> Shuffle</h4>
<p><code>bed_shuffle</code> shuffles input intervals. Interval sizes are equal in the input and output.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y &lt;-<span class="st"> </span><span class="kw">bed_shuffle</span>(x)

<span class="kw">any</span>(x$start ==<span class="st"> </span>y$start)

<span class="kw">all</span>(x$end -<span class="st"> </span>x$start ==<span class="st"> </span>y$end -<span class="st"> </span>y$start)</code></pre></div>
</div>
</div>
<div id="other-topics" class="section level3">
<h3><span class="header-section-number">2.3.4</span> Other topics</h3>
<div id="grouping-related-intervals" class="section level4">
<h4><span class="header-section-number">2.3.4.1</span> Grouping related intervals</h4>
<p>The <code>group_by</code> operation from <code>dplyr</code> serves a similar purpose to BEDtools <code>groupby</code>, and can be combined with <code>dplyr::summarize</code> to calculate summary statistics from groups.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome)
y &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome)

<span class="kw">bed_intersect</span>(x, y) %&gt;%
<span class="st">  </span><span class="kw">group_by</span>(chrom) %&gt;%
<span class="st">  </span><span class="kw">summarize</span>(<span class="dt">overlap.total =</span> <span class="kw">sum</span>(.overlap))
<span class="co">#&gt; # A tibble: 25 × 2</span>
<span class="co">#&gt;   chrom overlap.total</span>
<span class="co">#&gt;   &lt;chr&gt;         &lt;int&gt;</span>
<span class="co">#&gt; 1  chr1      26156419</span>
<span class="co">#&gt; 2 chr10      13838133</span>
<span class="co">#&gt; 3 chr11      14005918</span>
<span class="co">#&gt; 4 chr12      14216949</span>
<span class="co">#&gt; 5 chr13      11850603</span>
<span class="co">#&gt; 6 chr14      11098228</span>
<span class="co">#&gt; # ... with 19 more rows</span></code></pre></div>
</div>
</div>
<div id="interval-statistics" class="section level3">
<h3><span class="header-section-number">2.3.5</span> Interval Statistics</h3>
<p><code>valr</code> provides several methods to assess statistical properties of interval sets including:</p>
<ul>
<li><code>bed_fisher()</code>: measure overlap significance of two sets of intervals.</li>
<li><code>bed_absdist()</code>: XXX</li>
<li><code>bed_reldist()</code>: quantify relative distances between query intervals and closest reference intervals</li>
<li><code>bed_jaccard()</code>: quantify extent of overlap between two sets of intervals</li>
<li><code>bed_projection()</code>: XXX</li>
</ul>
<p>Several of these methods were described in the <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529">Genometricorr</a> software package.</p>
<div id="fishers-test" class="section level4">
<h4><span class="header-section-number">2.3.5.1</span> Fisher’s test</h4>
<p>The Fisher’s test assesses whether two sets of intervals are drawn from the same background genome.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome)
y &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome)

<span class="kw">bed_fisher</span>(x, y, genome)
<span class="co">#&gt;   estimate p.value conf.low conf.high                             method</span>
<span class="co">#&gt; 1     2.38       0     2.36      2.39 Fisher&#39;s Exact Test for Count Data</span>
<span class="co">#&gt;   alternative</span>
<span class="co">#&gt; 1   two.sided</span></code></pre></div>
</div>
<div id="absolute-distance" class="section level4">
<h4><span class="header-section-number">2.3.5.2</span> Absolute distance</h4>
<p><code>bed_absdist()</code> - XXX</p>
</div>
<div id="relative-distance" class="section level4">
<h4><span class="header-section-number">2.3.5.3</span> Relative distance</h4>
<p><code>bed_reldist()</code> computes the relative distance between a query interval and the two closest reference intervals. The relative distance can range between <code>[0,0.5]</code>. If the overall distribution of distances is biased toward zero, then the query and reference intervals tend to be closer in space (<a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002529">17</a>). By default a summary with the frequency of each relative distance is reported.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome)
y &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome)

<span class="kw">bed_reldist</span>(x, y)
<span class="co">#&gt; # A tibble: 51 × 4</span>
<span class="co">#&gt;   reldist counts  total   freq</span>
<span class="co">#&gt;     &lt;dbl&gt;  &lt;int&gt;  &lt;int&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt; 1    0.00  19813 999955 0.0198</span>
<span class="co">#&gt; 2    0.01  20140 999955 0.0201</span>
<span class="co">#&gt; 3    0.02  19658 999955 0.0197</span>
<span class="co">#&gt; 4    0.03  19790 999955 0.0198</span>
<span class="co">#&gt; 5    0.04  20032 999955 0.0200</span>
<span class="co">#&gt; 6    0.05  20389 999955 0.0204</span>
<span class="co">#&gt; # ... with 45 more rows</span></code></pre></div>
</div>
<div id="jaccard-similarity" class="section level4">
<h4><span class="header-section-number">2.3.5.4</span> Jaccard similarity</h4>
<p><code>bed_jaccard()</code> quantifies the extent of overlap between to sets of intervals. The Jaccard statistic takes values of <code>[0,1]</code> and is measured as:</p>
<p><span class="math display">\[
J(x,y) = \frac{\mid x \bigcap y \mid}
              {\mid x \bigcup y \mid} 
       = 
         \frac{\mid x \bigcap y \mid}
              {\mid x \mid + \mid y \mid - \mid x \bigcap y \mid}
\]</span></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">x &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome)
y &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome)

<span class="kw">bed_jaccard</span>(x, y)
<span class="co">#&gt; # A tibble: 1 × 4</span>
<span class="co">#&gt;       len_i      len_u jaccard      n</span>
<span class="co">#&gt;       &lt;int&gt;      &lt;int&gt;   &lt;dbl&gt;  &lt;int&gt;</span>
<span class="co">#&gt; 1 322727018 2000000000   0.192 645541</span></code></pre></div>
</div>
<div id="projection" class="section level4">
<h4><span class="header-section-number">2.3.5.5</span> Projection</h4>
<p><code>bed_projection()</code> TBD</p>
</div>
</div>
<div id="benchmarks" class="section level3">
<h3><span class="header-section-number">2.3.6</span> Benchmarks</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># two sets of 1e6 random 1 kb intervals from hg19</span>
x &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome)
y &lt;-<span class="st"> </span><span class="kw">bed_random</span>(genome)

<span class="kw">library</span>(microbenchmark)
<span class="kw">microbenchmark</span>(
  <span class="kw">bed_random</span>(genome),
  <span class="kw">bed_closest</span>(x, y),
  <span class="kw">bed_intersect</span>(x, y),
  <span class="kw">bed_merge</span>(x),
  <span class="kw">bed_subtract</span>(x, y),
  <span class="kw">bed_complement</span>(x, genome),
  <span class="dt">times =</span> <span class="dv">1</span>,
  <span class="dt">unit =</span> <span class="st">&#39;s&#39;</span>
)
<span class="co">#&gt; Unit: seconds</span>
<span class="co">#&gt;                       expr   min    lq  mean median    uq   max neval</span>
<span class="co">#&gt;         bed_random(genome) 0.232 0.232 0.232  0.232 0.232 0.232     1</span>
<span class="co">#&gt;          bed_closest(x, y) 8.905 8.905 8.905  8.905 8.905 8.905     1</span>
<span class="co">#&gt;        bed_intersect(x, y) 3.106 3.106 3.106  3.106 3.106 3.106     1</span>
<span class="co">#&gt;               bed_merge(x) 4.476 4.476 4.476  4.476 4.476 4.476     1</span>
<span class="co">#&gt;         bed_subtract(x, y) 2.084 2.084 2.084  2.084 2.084 2.084     1</span>
<span class="co">#&gt;  bed_complement(x, genome) 4.211 4.211 4.211  4.211 4.211 4.211     1</span></code></pre></div>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="demos.html" class="navigation navigation-next " aria-label="Next page""><i class="fa fa-angle-right"></i></a>

<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/jayhesselberth/valr-demo/edit/master/functions.Rmd",
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
