<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>valr: Genome interval arithmetic in R</title>
  <meta content="text/html; charset=UTF-8" http-equiv="Content-Type">
  <meta name="description" content="This book documents usage and examples of the valr package for genome interval arithmetic.">
  <meta name="generator" content="bookdown 0.1.7 and GitBook 2.6.7">

  <meta property="og:title" content="valr: Genome interval arithmetic in R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="This book documents usage and examples of the valr package for genome interval arithmetic." />
  <meta name="github-repo" content="jayhesselberth/valr-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="valr: Genome interval arithmetic in R" />
  <meta name="twitter:site" content="@jayhesselberth" />
  <meta name="twitter:description" content="This book documents usage and examples of the valr package for genome interval arithmetic." />
  

<meta name="author" content="Jay Hesselberth">

<meta name="date" content="2016-09-12">

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  

<link rel="next" href="functions.html">

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>


  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Genome interval arithmetic in R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#install"><i class="fa fa-check"></i><b>1.1</b> Installation</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#tool-compare"><i class="fa fa-check"></i><b>1.2</b> Comparison to other tools</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#contribute"><i class="fa fa-check"></i><b>1.3</b> Contribute</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="functions.html"><a href="functions.html"><i class="fa fa-check"></i><b>2</b> Functions</a><ul>
<li class="chapter" data-level="2.1" data-path="functions.html"><a href="functions.html#non-standard-evaluation"><i class="fa fa-check"></i><b>2.1</b> Non-standard evaluation</a></li>
<li class="chapter" data-level="2.2" data-path="functions.html"><a href="functions.html#reading-data"><i class="fa fa-check"></i><b>2.2</b> Reading data</a><ul>
<li class="chapter" data-level="2.2.1" data-path="functions.html"><a href="functions.html#bed-files"><i class="fa fa-check"></i><b>2.2.1</b> BED files</a></li>
<li class="chapter" data-level="2.2.2" data-path="functions.html"><a href="functions.html#genome-files"><i class="fa fa-check"></i><b>2.2.2</b> Genome files</a></li>
<li class="chapter" data-level="2.2.3" data-path="functions.html"><a href="functions.html#bedgraph-files"><i class="fa fa-check"></i><b>2.2.3</b> bedGraph files</a></li>
<li class="chapter" data-level="2.2.4" data-path="functions.html"><a href="functions.html#vcf-files"><i class="fa fa-check"></i><b>2.2.4</b> VCF files</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="functions.html"><a href="functions.html#interval-manipulations"><i class="fa fa-check"></i><b>2.3</b> Interval manipulations</a><ul>
<li class="chapter" data-level="2.3.1" data-path="functions.html"><a href="functions.html#single-set-operations"><i class="fa fa-check"></i><b>2.3.1</b> Single set operations</a></li>
<li class="chapter" data-level="2.3.2" data-path="functions.html"><a href="functions.html#multiple-set-operations"><i class="fa fa-check"></i><b>2.3.2</b> Multiple set operations</a></li>
<li class="chapter" data-level="2.3.3" data-path="functions.html"><a href="functions.html#randomzing-intervals"><i class="fa fa-check"></i><b>2.3.3</b> Randomzing intervals</a></li>
<li class="chapter" data-level="2.3.4" data-path="functions.html"><a href="functions.html#other-topics"><i class="fa fa-check"></i><b>2.3.4</b> Other topics</a></li>
<li class="chapter" data-level="2.3.5" data-path="functions.html"><a href="functions.html#interval-statistics"><i class="fa fa-check"></i><b>2.3.5</b> Interval Statistics</a></li>
<li class="chapter" data-level="2.3.6" data-path="functions.html"><a href="functions.html#benchmarks"><i class="fa fa-check"></i><b>2.3.6</b> Benchmarks</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="demos.html"><a href="demos.html"><i class="fa fa-check"></i><b>3</b> Demonstrations</a><ul>
<li class="chapter" data-level="3.1" data-path="demos.html"><a href="demos.html#bedtools-dnase-i"><i class="fa fa-check"></i><b>3.1</b> BEDtools DNase I</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="functions.html"><a href="functions.html#benchmarks"><i class="fa fa-check"></i><b>4</b> Benchmarks</a><ul>
<li class="chapter" data-level="4.1" data-path="benchmarks.html"><a href="benchmarks.html"><i class="fa fa-check"></i><b>4.1</b> Intersection</a></li>
<li class="chapter" data-level="4.2" data-path="benchmarks.html"><a href="benchmarks.html#closest-1"><i class="fa fa-check"></i><b>4.2</b> Closest</a></li>
<li class="chapter" data-level="4.3" data-path="benchmarks.html"><a href="benchmarks.html#merge-1"><i class="fa fa-check"></i><b>4.3</b> Merge</a></li>
<li class="chapter" data-level="4.4" data-path="benchmarks.html"><a href="benchmarks.html#subtract-1"><i class="fa fa-check"></i><b>4.4</b> Subtract</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">valr: Genome interval arithmetic in R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">valr: Genome interval arithmetic in R</h1>
<h4 class="author"><em>Jay Hesselberth</em></h4>
<h4 class="date"><em>2016-09-12</em></h4>
</div>
<div id="intro" class="section level1">
<h1><span class="header-section-number">Chapter 1</span> Introduction</h1>
<p><code>valr</code> provides tools to read and manipulate genome intervals and signals, similar to the standalone <a href="http://bedtools.readthedocs.org/en/latest/"><code>BEDtools</code></a> suite. <code>valr</code> enables BEDtools-like analysis in the R/RStudio environment, and uses modern R tools for a terse, expressive syntax. Compute-intensive algorithms are implemented in <a href="http://www.rcpp.org/"><code>Rcpp</code></a>/C++, and many methods take advantage of the speed and grouping capability provided by <a href="https://github.com/hadley/dplyr"><code>dplyr</code></a>.</p>
<p>We provide several introductions to <code>valr</code>:</p>
<ul>
<li>This comprehensive vignette that covers the core methods.</li>
<li>A tutorial that demonstrates how to use <code>valr</code> in “real-world” applications.</li>
<li>A <code>shiny</code> application that demonstrates interactive analysis of genome-scale data sets.</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(valr)
<span class="kw">library</span>(dplyr)</code></pre></div>
<div id="install" class="section level2">
<h2><span class="header-section-number">1.1</span> Installation</h2>
<p><code>valr</code> can be installed from github, and will be eventually deposited in CRAN.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools::<span class="kw">install_github</span>(<span class="st">&#39;jayhesselberth/valr&#39;</span>)</code></pre></div>
</div>
<div id="tool-compare" class="section level2">
<h2><span class="header-section-number">1.2</span> Comparison to other tools</h2>
<p><strong>Why another tool set for interval manipulations?</strong> We already have <a href="http://bedtools.readthedocs.org/en/latest/">BEDtools</a>, <a href="http://bedops.readthedocs.org/en/latest/index.html">bedops</a>, <a href="https://pythonhosted.org/pybedtools/">pybedtools</a>, <a href="https://bioconductor.org/packages/release/bioc/html/GenomicRanges.html">GenomeRanges</a>, <a href="https://cran.r-project.org/web/packages/bedr/index.html">bedr</a> and <a href="https://bioconductor.org/packages/release/bioc/html/IRanges.html">IRanges</a>.</p>
<p>We were motivated to develop a toolset that:</p>
<ul>
<li>Combines analysis and visualization in RStudio.</li>
<li>Can be used to generate reports with Rmarkdown.</li>
<li>Is highly extensible. New tools are quickly implemented on the R side.</li>
<li>Leverages the “modern R” syntax, using <code>dplyr</code> and the pipe operator from <code>magrittr</code> (<code>%&gt;%</code>).</li>
<li>Maximizes speed by implementing compute-intensive algorithms in <code>Rcpp</code>.</li>
<li>Facilitates interactive visulaizations with <a href="http://shiny.rstudio.com/"><code>shiny</code></a>.</li>
</ul>
<p>We anticipate <code>valr</code> will mainly be used for analysis of pre-processed data in BED, bedGraph and VCF formats. Most users will have processed their aligned reads from BAM format to bedGraph, so we do not foresee supporting BAM directly. We would entertain requests for GTF / GFF support if there is interest, as <code>tidyr</code> makes it easy to convert these to BED12.</p>
<p>Certain algorithms in <code>valr</code> were implemented in <code>Rcpp</code> (including <a href="https://github.com/jayhesselberth/valr/tree/master/src">intersect, merge, subtract, closest</a>) to enable fluid interactive analysis. See the <a href="functions.html#benchmarks">benchmarks</a> section for details.</p>
</div>
<div id="contribute" class="section level2">
<h2><span class="header-section-number">1.3</span> Contribute</h2>
<p><code>valr</code> is actively maintained on [github][<a href="http://github.com/jayhesselberth/valr" class="uri">http://github.com/jayhesselberth/valr</a>] and we welcome ideas for new tools and pull requests.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="functions.html" class="navigation navigation-next navigation-unique" aria-label="Next page""><i class="fa fa-angle-right"></i></a>

<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/jayhesselberth/valr-demo/edit/master/index.Rmd",
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    if (location.protocol !== "file:" && /^https?:/.test(script.src))
      script.src  = script.src.replace(/^https?:/, '');
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
